### PR TITLE
[AST] NFC: Simplify AnyFunctionRef::isKnownNoEscape()

### PR DESCRIPTION
**Pull request by davezarzycki**
This removes a defer body hack and an assumption that non-defer functions are always escaping.

### CODE REVIEW

**Comment by davezarzycki** [comment]
@swift-ci please test

**Comment by CodaFi** [comment]
This doesn’t remove it so much as shuffle it into the type checker.  Can you document that?

**Comment by davezarzycki** [comment]
Sure. For my own edification and a better comment in the code, why is it still a hack if the type is set correctly from the start?

**Comment by CodaFi** [comment]
>why is it still a hack if the type is set correctly from the start?

It’s a hack, but less of a hack now that it’s in`configureInterfaceType` which has become Too Big To Fail and needs to be simplified and broken up.  One way or another I have a working theory that there’s a better way to do 90% of the things in there elsewhere.

**Comment by davezarzycki** [comment]
I think I'm missing something. What you wrote seems like general feedback about `configureInterfaceType` and not specific feedback about my change. As you requested, I'd like to document this pull request. Can you please be more specific about how/where you want documentation added? Thanks

File: lib/Sema/TypeCheckGeneric.cpp [hunk 1]
```diff
@@ -873,8 +873,11 @@ void TypeChecker::configureInterfaceType(AbstractFunctionDecl *func,
 
     // 'throws' only applies to the innermost function.
     AnyFunctionType::ExtInfo info;
-    if (i == 0 && func->hasThrows())
-      info = info.withThrows();
+    if (i == 0) {
+      info = info.withThrows(func->hasThrows());
+      if (auto fd = dyn_cast<FuncDecl>(func))
```
Comment (CodaFi) [inline lib/Sema/TypeCheckGeneric.cpp]: Document this by saying (roughly) “defer bodies are always noescape”

**Comment by davezarzycki** [comment]
Thanks for the feedback @CodaFi.
@swift-ci please smoke test and merge.

**Comment by davezarzycki** [comment]
@swift-ci please smoke test and merge

**Comment by atrick** [comment]
@davezarzycki I'm not sure what your interest in `isKnownNoEscape` is, but FYI I've got a branch that lowers all function types not declared `@escaping` as `@noescape` SILFunctionTypes. That's going to enable stack allocation of closures. I  just need to work through a long list of test failures before pushing it.

**Comment by davezarzycki** [comment]
Hi @atrick! That's cool! For whatever it may be worth, I was doing some hacking/experimentation that tripped over `isKnownNoEscape` assuming that all function decls (other than defer statements) were `@escaping`. I figured I might as well push the cleanup upstream. :-)

**Pull request merged**